wasi/wasi.go

```diff
@@ -100,88 +170,102 @@ type fileEntry struct {
 	file    File
 }
 
-type Option func(*WASIEnvironment)
+type Option func(*api)
 
 func Stdin(reader io.Reader) Option {
-	return func(w *WASIEnvironment) {
+	return func(w *api) {
 		w.stdin = reader
 	}
 }
 
 func Stdout(writer io.Writer) Option {
-	return func(w *WASIEnvironment) {
+	return func(w *api) {
 		w.stdout = writer
 	}
 }
 
 func Stderr(writer io.Writer) Option {
-	return func(w *WASIEnvironment) {
+	return func(w *api) {
 		w.stderr = writer
 	}
 }
 
-// wasiStringArray holds null-terminated strings. It ensures that
+// nullTerminatedStrings holds null-terminated strings. It ensures that
 // its length and total buffer size don't exceed the max of uint32.
 // Each string can have arbitrary byte values, not only utf-8 encoded text.
-// wasiStringArray are convenience struct for args_get and environ_get. (environ_get is not implemented yet)
+// nullTerminatedStrings are convenience struct for args_get and environ_get. (environ_get is not implemented yet)
 //
 // A Null-terminated string is a byte string with a NULL suffix ("\x00").
 // See https://en.wikipedia.org/wiki/Null-terminated_string
-type wasiStringArray struct {
+type nullTerminatedStrings struct {
 	// nullTerminatedValues are null-terminated values with a NULL suffix.
 	// Each string can have arbitrary byte values, not only utf-8 encoded text.
 	nullTerminatedValues [][]byte
 	totalBufSize         uint32
 }
 
-// newWASIStringArray creates a wasiStringArray from the given string slice. It returns an error
-// if the length or the total buffer size of the result WASIStringArray exceeds the max of uint32
-func newWASIStringArray(args []string) (*wasiStringArray, error) {
-	if args == nil {
-		return &wasiStringArray{nullTerminatedValues: [][]byte{}}, nil
+// newNullTerminatedStrings creates a nullTerminatedStrings from the given string slice. It returns an error
+// if the length or the total buffer size of the result WASIStringArray exceeds the maxBufSize
+func newNullTerminatedStrings(maxBufSize uint32, args ...string) (*nullTerminatedStrings, error) {
+	if len(args) == 0 {
+		return &nullTerminatedStrings{nullTerminatedValues: [][]byte{}}, nil
 	}
-	if len(args) > math.MaxUint32 {
-		return nil, fmt.Errorf("the length of the args exceeds the max of uint32: %v", len(args))
-	}
-	strings := make([][]byte, len(args))
+	var strings [][]byte // don't pre-allocate as this function is size bound
 	totalBufSize := uint32(0)
 	for i, arg := range args {
-		argLen := uint64(len(arg)) + 1 // + 1 for "\x00"
-		if argLen > uint64(math.MaxUint32-totalBufSize) {
-			return nil, fmt.Errorf("the required buffer size for the args exceeds the max of uint32: %v", uint64(totalBufSize)+argLen)
+		if !utf8.ValidString(arg) {
+			return nil, fmt.Errorf("arg[%d] is not a valid UTF-8 string", i)
+		}
+		argLen := uint64(len(arg)) + 1 // + 1 for "\x00"; uint64 in case this one arg is huge
+		nextSize := uint64(totalBufSize) + argLen
+		if nextSize > uint64(maxBufSize) { //
```

Comment:
looks like you're dying while writing :D

Reply:
lol

Reply:
I moved part of the comment up and left this..
```suggestion
		if nextSize > uint64(maxBufSize) {
```